### DISCUSSION TITLE
package-scoped automation tokens for publishing

### DISCUSSION THREAD

**Discussion by mxschmitt** [npm/feedback#457] [Registry] [answered]
When maintaining a lot of different projects on NPM, it's a security risk, to not be able to scope a NPM automation API key to a specific project. When for one project the access rules gets bypassed on e.g. GitHub e.g. when a maintainers account is compromised, he can use the GitHub Action secret to also publish new malicious releases on different completely unrelated projects which belong to the user who created the API key.

This is exactly what https://github.com/npm/roadmap/issues/10 proposes. But since we can't comment or vote on the roadmap, I created this discussion here.

**Comment by MylesBorins** [comment, marked as answer]
@mxschmitt anything on the roadmap is work that we've already committed to and opening additional feedback issues does create additional noise and work for the team, so please refrain from doing this in the future

**Comment by broofa** [comment]
Landed here intending to ask for this same feature (package-scoped access tokens.)

My use case is pretty simple: I maintain a couple top-50 modules (`mime`, `uuid`).  Combined, these are downloaded ~110M times/week.  Very popular.  Very important that they don't get compromised.

The problem is, if I need to create an access token for some trivial purpose - such as setting up automated releases for my `runmd` project, which is what I'm currently working on - that token becomes an attack surface for all of these modules.  As a result, I find myself reluctant to use access tokens.  I need to use them, of course, but the increased risk each token represents _to projects I'm not actively working on_ definitely has my spidey-security-sense tingling.

@MylesBorins I know there's probably not a lot of new insight in this, but I thought this case might factor into how you prioritize this feature.  Speaking of which, I'm also commenting because a year ago you said this:

> anything on the roadmap is work that we've already committed to 

... but then in March you said [this](https://github.com/npm/roadmap/issues/10#issuecomment-1068057128):

> We are continuing to scope this feature and will add it to the GitHub public roadmap when we the work is funded and we have a target date.

... which has a distinctly, "not actually on the roadmap" vibe to it. 😞

**Reply by MylesBorins** [reply]
@broofa I can assure you we are definitely working on scoping what we are currently calling "granular access tokens". The only reason it isn't on the roadmap yet is because we are still working through the early engineering ADR + design work. Until we understand full cost + scope we can't accurately determine when we would ship it. We want to avoid making promises about timelines prior to knowing these things.

@monishcm is driving the product for this and we'll hopefully have an update for folks in the near future.
